docs/GDPs/gdp-0002.rst

```diff
@@ -0,0 +1,365 @@
+It is the responsibility of this object to translate from a global index provided by the user to a local index used by gt4py.
+The processes that do not have an global iteration space index intersection with the provided regions simply skip this computation altogether.
+
+
+Usage and Impact
+----------------
+
+This is an optional feature, but should help nearly every application of GT4Py, since almost all partial differential equations at some point have a non-periodic boundary condition or require special treatment somewhere in the domain.
+
+
+Backward Compatibility
+----------------------
+
+This GDP aims to be fully backward-compatible.
+
+
+Detailed Description
+--------------------
+
+The intent of this GDP is to a specify horizontal bounds inside the stencil in a similar fashion to how vertical intervals are specified.
+However, the fact that there are multiple dimensions in the horizontal, halos, and domain decomposition imply the feature needs to be more complex than a simple index range used for the vertical.
+
+As introduced above, we propose adding a new gtscript `region()` specifier that specializes the stencil on a region of the space when a non-zero intersection with the global region occurs.
+The idea behind using the region objects is that they can incorporate both the iteration space a process sees and the global partitioning that an application must handle, all while hiding the complexity at the stencil level for model developers.
+In this proposal, the application remains in control of the partitioning data so **the partitioner is entirely transparent to both GT4Py and its backend.**
+
+
+Specification Options
++++++++++++++++++++++
+
+An issue we faced when designing the iteration space featiure is that the stages (`with interval()`) are, by the execution model specification, free to be re-ordered by the runtime.
+We propose adding the `region()` to **either** (1) the multistage, in which case the execution order is specified by GridTool's execution model **or** (2) within a stage.
+We plan to implement these in order, first finishing an implementation of (1) before attempting the second.
+If we agree to try (2) the example would look like:
+
+.. code-block:: python
+
+    @gtscript.stencil(backend=backend_type)
+    def stencil(uc: FIELD_TYPE, vc: FIELD_TYPE, cosa: FIELD_TYPE, rsina: FIELD_TYPE, ut: FIELD_TYPE, ub: FIELD_TYPE, dt5: float, dt4: float):
+        with computation(PARALLEL), interval(...):
+            ub = dt5 * (uc[0, -1, 0] + uc - (vc[-1, 0, 0] + vc) * cosa) * rsina
+            with region(WestEdge(0, 1), EastEdge(0, 1)):
+                ub = dt5 * (ut[0, -1, 0] + ut)
+            with region(SouthEdge(0, 1), NorthEdge(0, 1)):
+                ub = dt4 * (-ut[0, -2, 0] + 3.0 * (ut[0, -1, 0] + ut) - ut[0, 1, 0])
```

Comment:
If we were to extend this example to multiple vertical regions, would it look like this:?
```
with computation(PARALLEL), interval(...):
            ub = dt5 * (uc[0, -1, 0] + uc - (vc[-1, 0, 0] + vc) * cosa) * rsina
            with region(WestEdge(0, 1), EastEdge(0, 1)):
                ub = dt5 * (ut[0, -1, 0] + ut)
            with region(SouthEdge(0, 1), NorthEdge(0, 1)):
                ub = dt4 * (-ut[0, -2, 0] + 3.0 * (ut[0, -1, 0] + ut) - ut[0, 1, 0])
 with computation(PARALLEL), interval(another_interval):
            ub = dt5 * (uc[0, -1, 0] + uc - (vc[-1, 0, 0] + vc) * cosa) * rsina
            with region(WestEdge(0, 1), EastEdge(0, 1)):
                ub = dt5 * (ut[0, -1, 0] + ut)
            with region(SouthEdge(0, 1), NorthEdge(0, 1)):
                ub = dt4 * (-ut[0, -2, 0] + 3.0 * (ut[0, -1, 0] + ut) - ut[0, 1, 0])
 ```

Reply:
Yes! I imagined that it could be either a new computation or not:

```python
with computation(PARALLEL)
    with interval(...):
            ub = dt5 * (uc[0, -1, 0] + uc - (vc[-1, 0, 0] + vc) * cosa) * rsina
            with region(WestEdge(0, 1), EastEdge(0, 1)):
                ub = dt5 * (ut[0, -1, 0] + ut)
            with region(SouthEdge(0, 1), NorthEdge(0, 1)):
                ub = dt4 * (-ut[0, -2, 0] + 3.0 * (ut[0, -1, 0] + ut) - ut[0, 1, 0])
     with interval(another_interval):
            ub = dt5 * (uc[0, -1, 0] + uc - (vc[-1, 0, 0] + vc) * cosa) * rsina
            with region(WestEdge(0, 1), EastEdge(0, 1)):
                ub = dt5 * (ut[0, -1, 0] + ut)
            with region(SouthEdge(0, 1), NorthEdge(0, 1)):
                ub = dt4 * (-ut[0, -2, 0] + 3.0 * (ut[0, -1, 0] + ut) - ut[0, 1, 0])
```

 However, this version will not be supported in the first implementation of `region()`. The current first implementation allows the `region()` `withstmt` to be on the `computation()`, in which case it applies to all intervals()`, or `interval()`, in which case it applies to that interval only.

I would like to support a switch-like `region`, but I am cautious about how to map this to the GridTools backends. I was planning to use boundary conditions, which requires a new stencil call for each boundary. I'm not sure how this would map over to separate vertical regions. I suppose we could break the intervals into new multistages, then generate separate stencil calls for each (6 GPU kernel calls in the code I printed above).

For now we can focus on customized backends, but I would really like to support GridTools with good performance. Therefore, it would be fantastic if GridTools were to eventually support such a switch-like case within a stencil. I can see how it would lead to quite poor performance on CPUs, but potentially would be fine for GPUs. Is that a possibility for future versions? :)